imsim/diffraction.py

```diff
@@ -42,7 +42,61 @@ class Geometry:
 )
 
 
-def diffraction_kick(
+def apply_diffraction_delta(
+    pos: np.ndarray,
+    v: np.ndarray,
+    t: np.ndarray,
+    wavelength: np.ndarray,
+    lat: float,
+    telescope_orientation: np.ndarray,
+    geometry: Geometry,
+    distribution: Callable[[np.ndarray], np.ndarray],
+) -> np.ndarray:
+    """Statistically diffract photons.
+
+    For a ray of photons with positions pos, wavelengths wavelength and pupil plane
+    coordinates pos,
+    randomly generate diffraction angles and change the directions of the photons
+    entering the pupil plane.
+
+    Parameters
+    ----------
+    pos : 2d positions of the intersections of the rays with the pupil plane (shape (n, 2)).
+    v : Direction of the photons (shape (n, 3)).
+    t : Time [sec] when photons enter the pupil plane (shape (n,)).
+    wavelength : Wavelength of the photons (shape (n,)).
+    lat : Latitude of the telescope.
+    telescope_orientation : A np.ndarray as prepared by `telescope_orientation_matrix`.
+    geometry : Geometry representing the 2d projection of the spider of a telescope
+        into the pupil plane.
+    distribution : Random number generator representing the distribution of the diffraction
+        angles (depending on the wavelength and distance to the geometry).
+        It should be a callable accepting an array for phi_star and returning the
+        random values for phi
+    """
+    R = field_rotation_matrix(lat, (telescope_orientation @ v).T, t)
+
+    def rot(v):
+        return np.einsum("kij,kj->ki", R, v)
+
+    def rot_inv(v):
+        return np.einsum("kji,kj->ki", R, v)
```

Comment:
One more thing. (I apparently started writing this and forgot to hit submit.)
This strikes me as possibly a tall pole for this calculation. Josh did some work investigating einsum timings and found that dot is quite a bit faster. So if this is possible to reorder into something that works with dot, that would probably be worth doing.
cf. https://github.com/jmeyers314/SeriesNotebooks/blob/master/np.einsum%20vs%20np.dot%20speed.ipynb

Reply:
Yep, `(R @ pos[..., None]).squeeze(axis=-1)` would be possible (or `:, :` instead of `...`)

Reply:
... but:
```py
import numpy as np
n = 10000
A = np.random.rand(n, 2, 2)
x = np.random.rand(n, 2)
%timeit y = (A @ x[:, :, None]).squeeze(axis=-1)
# 371 µs ± 12.5 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
%timeit y = np.einsum("kij,kj->ki", A, x)
# 105 µs ± 2.31 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

Reply:
> Josh did some work investigating einsum timings and found that dot is quite a bit faster

Wow, that's a deep cut.  I completely forgot I did that :)  I wonder if it holds up after 7 years...

Reply:
lol.  You were among the first few hits on whatever google terms I searched for trying to see how einsum might compare to other options.  :)